src/eventmodel.ts

```diff
@@ -0,0 +1,580 @@
+//   3. Listen on a subtree and fire when anything in the subtree changes. This
+//   is used for custom functions.
+//
+// {{foo.id}} would listen on the fixed path ['foo', 'id'].
+//
+//
+// - Context tree represents the changing (embedded) contexts of the templating
+//   engine. This maps to the tree of templates and allows templates to reference
+//   anything in any of their enclosing template scopes.
+//
+
+// module.exports = EventModel;
+
+// The code here uses object-based set pattern where objects are keyed using
+// sequentially generated IDs.
+let nextId = 1;
+
+// A binding object is something with update(), insert()/move()/remove() defined.
+
+// Given x[y] with model.get(y) == 5:
+//  item = 5
+//  segments = ['y']
+//  outside = the EventModel for x.
+//
+// Note that item could be a Context or another ModelRef - eg:
+//
+// {{ each foo as bar }} ... {{ x[bar] }}  -or-  {{ x[y[z]] }}
+class ModelRef{
+  id: number;
+  model: Model;
+  segments: string[];
+  item: any;
+  outside: any;
+  result: any;
+  
+  constructor(model, item, segments, outside) {
+    this.id = nextId++;
+    // We need a reference to the model & our segment list so we can update our
+    // value.
+    this.model = model;
+    this.segments = segments;
+    // Our current value.
+    this.item = item;
+    // outside is a reference to the EventModel of the thing on the lhs of the
+    // brackets. For example, in x[y].z, outside is the EventModel of x.
+    this.outside = outside;
+
+    // result is the EventModel of the evaluated version of the brackets. In
+    // x[y].z, its the EventModel of x[y].
+    this.result = outside.child(item).refChild(this);
+  }
+  
+  update() {
+    const segments = expressions.pathSegments(this.segments);
+    const newItem = expressions.lookup(segments, this.model.data);
+    if (this.item === newItem) return;
+  
+    // First remove myself.
+    delete this.outside.child(this.item).refChildren[this.id];
+  
+    this.item = newItem;
+  
+    const container = this.outside.child(this.item);
+    // I want to just call refChild but that would create a new EM. Instead I
+    // want to just implant my current EM there.
+    if (!container.refChildren) container.refChildren = new RefChildrenMap();
+    container.refChildren[this.id] = this.result;
+  
+    // Finally, update all the bindings in the tree.
+    this.result.update();
+  }
+}
+
+
+class RefOutMap {}
+class RefChildrenMap {}
+class BindingsMap {}
+class ItemContextsMap {}
+class EventModelsMap {}
+
+function hasKeys(object) {
+  for (const key in object) {
+    return true;
+  }
+  return false;
+}
+ 
+function childSetWildcard(child) {
+  child._set();
+}
+
+export class EventModel {
+  array: any;
+  arrayByReference: any;
+  bindings: BindingsMap | null;
+  dependancies: any;
+  id: number;
+  itemContexts: ItemContextsMap | null;
+  object: any;
+  refChildren: RefChildrenMap | null;
+  refOut: RefOutMap | null;
+
+  cosntructor() {
```

Comment:
Typo here is causing the failing tests. The constructor is strictly speaking only necessary for setting the EventModel's id, which seems to only matter for specific kinds of binding changes.

```suggestion
  constructor() {
```

Debugging process:

1. Compared the invocations of the test's `len()` function between the original JS version and this TS version. With this TS version, the stack traces were the same minus line number drift, but it had two extra invocations at the end.
2. Hopping up to the `EventModel.localUpdate` stack frame, I noticed the EventModel's `bindings` map had two bindings in the TS version, versus one binding in the JS version.
3. Poked around a bit more, found that the constructor wasn't getting invoked in the TS version. Stepping into the `new EventModel()` just pointed to an empty line in eventmodel.js.
4. Puzzled, I switched tests over to use prebuilt code in `dist`, and that was when I saw that the JS version had a separately defined `cosntructor` function, instead of the constructor body being in the EventModel function body.

<img width="428" alt="Screenshot 2024-07-02 at 4 49 14 PM" src="https://github.com/derbyjs/derby/assets/6915076/226ccd10-e5b1-4a44-b274-30a2fc58bd6b">